### ISSUE TITLE
disable selector caching by default (only cache a single value)

### ISSUE THREAD

**Issue by ahfarmer** [facebookexperimental/Recoil#1064]
I've been diving deep into some recoil-related memory leaks on my site and I was shocked to discover that `selector` caches EVERY previous value it has ever held.

I see there are some new caching policies in the works, but I think this 'cache everything' policy should not be the default. There is no mention of this caching in the docs that I could find.

This page doesn't mention any caching: https://recoiljs.org/docs/basic-tutorial/selectors
In fact that page explicitly says that the selectors will re-run when the inputs changed, which is misleading.

Of course if a selector is fetched with `useRecoilValue()` it shouldn't be recomputed every time. But I believe that by default a selector should only cache ONE value. If the inputs change, they should throw away all previous values and just keep the newly computed value.

**Comment by ahfarmer** [comment]
To expand on this a bit more, I would expect that I could clear a selector by clearing its upstream atoms.

For example, lets say I have a counter atom and a selector that creates a very large 60MB object downstream of that atom:

```javascript
const counterState = atom({
  key: "counterState",
  default: null,
});

// simulate creating an object that depends on the counter and takes up a lot of memory
const counterArrayState = selector({
  key: "counterArrayState",
  get: ({ get }) => {
    const counter = get(counterState);
    if (!counter) {
      return null;
    }
    let a = [];
    for (let i = 0; i < ARRAY_SIZE; i++) {
      a.push(counter);
    }
    return a;
  },
  // cachePolicy_UNSTABLE: { eviction: "lru", maxSize: 1 },
});
```

I would expect that if I set counterState to null, and then get() this selector, it should recalculate as null and give me back my 60MB of memory. Unfortunately, by default it will hold onto that memory forever. With the LRU cache policy it looks like it holds a minimum of 2-3 elements (even with maxSize set to 1).

**Comment by davidmccabe** [comment]
Sorry about the surprise, this isn't a good default and the docs could be clearer as you pointed out.

We're planning to roll out configuration options for selector caching in the next release or so. We previously had these options but had to change them because they were coupled to the specific way the implementation worked.

@csantos42 FYI

**Comment by ahfarmer** [comment]
May I suggest that before you add the caching options you disable this
extreme level of undocumented caching?

I already switched to Jotai because with Recoil my app was leaking 10 megs
a minute. No leaks with jotai and the code is very similar.

**Comment by drarmstr** [comment]
[Configurable selector caches](https://recoiljs.org/blog/2021/07/30/0.4.0-release#configurable-selector-caches) now available in Recoil 0.4.

**Comment by cypresslf** [comment]
I was bit by the defaults as well. It would be nice to have a saner default, not just an optional configuration
